src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -168,3 +185,316 @@ RegisterTenantSchema(Oid schemaId)
+	char *schemaName = get_namespace_name(schemaId);
+	EnsureTenantSchemaAllowed(schemaId);
+	EnsureSchemaOwner(schemaId);
+	EnsureCoordinator();
+
+	/* Prevent any update to the schema */
+	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
+
+	/* Return if the schema is already a tenant schema */
+	if (IsTenantSchema(schemaId))
+	{
+		ereport(NOTICE, (errmsg("schema %s is already a tenant schema", schemaName)));
+		PG_RETURN_VOID();
+	}
+	else
+	{
+		ereport(NOTICE, (errmsg("schema %s is converted to a tenant schema",
+								schemaName)));
+	}
+
+	/* Retrieve all tables in the schema */
+	bool hasForeignTable = false;
+	List *relationIdsInSchema = GetRelationIdsInSchema(schemaId, &hasForeignTable);
+	if (hasForeignTable)
+	{
+		ereport(ERROR, (errmsg("foreign tables are not allowed in a tenant schema")));
+	}
+
+	/*
+	 * There should not exist any distributed tables other than Citus local tables
+	 * inside the schema.
+	 */
+	if (HasAnyDistributedTableUnderSchema(relationIdsInSchema))
+	{
+		ereport(ERROR, (errmsg("schema already has distributed tables"),
+						errhint("Undistribute distributed tables before setting "
+								"a tenant schema.")));
+	}
+
+	/* First, register the schema and sync it to workers */
+	char *registerSchemaCommand = RegisterTenantSchema(schemaId);
+	if (EnableMetadataSync)
+	{
+		SendCommandToWorkersWithMetadata(registerSchemaCommand);
+	}
+
+	/* Convert all tables in the schema into single shard table */
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdsInSchema)
+	{
+		CreateTenantSchemaTable(relationId);
+	}
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * citus_schema_tenant_unset gets a tenant schema name, then converts it to a regular
+ * schema.
+ */
+Datum
+citus_schema_tenant_unset(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	EnsureCoordinator();
+
+	Oid schemaId = PG_GETARG_OID(0);
+	char *schemaName = get_namespace_name(schemaId);
+	EnsureTenantSchemaAllowed(schemaId);
+	EnsureSchemaOwner(schemaId);
+
+	/* Prevent any update to the schema */
+	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
+
+	/* All tables inside the schema should be single shard distributed */
+	if (!IsTenantSchema(schemaId))
+	{
+		ereport(ERROR, (errmsg("schema %s is not a tenant schema", schemaName)));
+	}
+
+	ereport(NOTICE, (errmsg("schema %s is converted back to a regular schema",
+							schemaName)));
+
+	/* store foreign key creation commands that relation is involved */
+	List *originalForeignKeyRecreationCommands = NIL;
+
+	/* Retrieve all tables in the schema */
+	bool hasForeignTable = false;
+	List *relationIdsInSchema = GetRelationIdsInSchema(schemaId, &hasForeignTable);
+	bool allSingleShardTable PG_USED_FOR_ASSERTS_ONLY =
+		AllSingleShardTableUnderSchema(relationIdsInSchema);
+	Assert(!hasForeignTable && allSingleShardTable);
+
+	/* Undistribute all tables in the schema */
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdsInSchema)
```

Comment:
Given that single shard tables are becoming a first-class citizen, I'm kind of wondering whether we really need to undistribute here. Seems like the user could decide to do that separately?

Reply:
I also wonder if we should only remove pg_dist_tenant_schema metadata here. And let users decide them to convert. This is also a plus for us to not block too much.

Reply:
See latest commit.

- I opt for not undistributing tables during unset tenant schema. I do not remove colocation id for the single shard tables in this case. 
- And during set tenant schema, I use already existing colocation id if there is any single shard table under the registered schema.
- We allow only Citus local tables, PG local tables, and colocated single shard tables during set tenant schema.
- We expect only single shard tables exist during unset tenant schema.

Reply:
I am still not sure about this. A typical use case for using unset is when we do not support some query for single shards, the user undistributes all table via unset. Maybe we should add an option like `undistribute` by default true. User can set it false if they do not want this behaviour.